### PR TITLE
[SPARK-29712][SQL] Take into account the left bound in `fromDayTimeString()`

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to check the left bound - the `from` interval unit, and reset all unit values out of the specified range of interval units. Currently, `IntervalUtils.fromDayTimeString()` takes into account only the right bound but not the left one. 

**Note**: the reset is not performed if `spark.sql.dialect = PostgreSQL`.

### Why are the changes needed?
This fix makes `fromDayTimeString()` consistent in bound handling.

### Does this PR introduce any user-facing change?
Yes, before:
```sql
spark-sql> SELECT interval '1 2:03:04' hour to minute;
interval 1 days 2 hours 3 minutes
```
After:
```sql
spark-sql> SELECT interval '1 2:03:04' hour to minute;
interval 2 hours 3 minutes
```

### How was this patch tested?
- Added new tests to `IntervalUtilsSuite`
- Fixed expected results for `literals.sql` and `interval.sql`

### CODE REVIEW

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by MaxGekk** [comment]
@srowen @dongjoon-hyun @cloud-fan Please, have a look at this.

**Comment by cloud-fan** [comment]
> the reset is not performed if spark.sql.dialect = PostgreSQL.

Do you mean pgsql has a different behavior?

**Comment by MaxGekk** [comment]
> Do you mean pgsql has a different behavior?

It ignores `from` but takes into account `to`. Not sure it is a bug or a feature.
```sql
maxim=# select interval '5 12:40:30.999999999' hour to minute;
    interval
-----------------
 5 days 12:40:00
(1 row)
```

**Comment by cloud-fan** [comment]
Does SQL spec define the behavior? I tried oracle and it fails if the from/to field doesn't match the actual value.

**Comment by srowen** [comment]
Pardon the dumb question, but what is the 'hour to minute' supposed to mean? select just the hours and minutes out of the interval? yeah, the pgsql implementation doesn't seem to do that, whether a bug or by design. Just trying to understand what we think the semantics are, and whether they need to be different in Spark.

**Comment by MaxGekk** [comment]
> what is the 'hour to minute' supposed to mean? select just the hours and minutes out of the interval?

Correct, it means to extract an interval with hour:minute from a string. 
Screen Shot 2019-11-04 at 18 52 14

**Comment by MaxGekk** [comment]
I haven't found precise description of acceptable input string but it seems the format should be defined by the interval qualifier. For example, if it is `HOUR to MINUTE`, the input string must be in the format `hh21:mi` but not any string as we suppose now: https://github.com/apache/spark/blob/a4382f7fe1c36a51c64f460c6cb91e93470e0825/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala#L182-L183

**Comment by srowen** [comment]
Hm, none of those examples show 'extracting' a subset of the interval. It seems like a way to specify what the unit-less string means. It seems like the pgsql you have above ought to fail if so, as the string doesn't match the units, yeah.

It seems like the current behavior works like pgsql? if so, is that maybe an OK stance on the semantics in these 'incorrect' cases?

**Comment by MaxGekk** [comment]
> if so, is that maybe an OK stance on the semantics in these 'incorrect' cases?

I am not sure that fully following other dbms in bugs and features is right decision. From the user point of view, if I specify `hour to minute`, I would expect only hours and minutes in the result column but not seconds and days, or something else.

**Comment by cloud-fan** [comment]
Since this is only for interval literal, I think it doesn't hurt if we fail invalid string format. The oracle behavior looks more reasonable to me, we can try more to see if that's a common behavior.

**Comment by MaxGekk** [comment]
For example, MySQL is strict as proposed in the PR:
```sql
mysql> SELECT DATE_SUB("1998-01-01 00:00:00", INTERVAL "1 1:1" HOUR_MINUTE);
+---------------------------------------------------------------+
| DATE_SUB("1998-01-01 00:00:00", INTERVAL "1 1:1" HOUR_MINUTE) |
+---------------------------------------------------------------+
| NULL                                                          |
+---------------------------------------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> SELECT DATE_SUB("1998-01-01 00:00:00", INTERVAL "1:1" HOUR_MINUTE);
+-------------------------------------------------------------+
| DATE_SUB("1998-01-01 00:00:00", INTERVAL "1:1" HOUR_MINUTE) |
+-------------------------------------------------------------+
| 1997-12-31 22:59:00                                         |
+-------------------------------------------------------------+
1 row in set (0.00 sec)
```

@cloud-fan @srowen What can I do to continue with the PR except of resolving the conflicts.

**Comment by srowen** [comment]
OK, so it would be reasonable to fail / return null, to match Oracle? There doesn't seem to be consistent behavior out there. I guess I prefer to fail this if it doesn't make sense semantically.

**Comment by cloud-fan** [comment]
so let's keep the existing behavior if pgsql dialect is set, and fail for invalid bounds otherwise.

**Comment by MaxGekk** [comment]
> so let's keep the existing behavior if pgsql dialect is set, and fail for invalid bounds otherwise.

This is what's implemented in the PR. Let me resolve conflicts.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/IntervalUtilsSuite.scala [hunk 1]
```diff
@@ -265,4 +267,30 @@ class IntervalUtilsSuite extends SparkFunSuite {
...
+  test("from bounded day-time string") {
+    val input = "5 12:40:30.999999999"
+
+    def check(from: IntervalUnit, to: IntervalUnit, expected: String): Unit = {
+      withClue(s"from = $from, to = $to") {
+        assert(fromDayTimeString(input, from, to) === fromString(expected))
+      }
+    }
+
+    withSQLConf(SQLConf.DIALECT.key -> "Spark") {
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/IntervalUtilsSuite.scala]: can we have a test case to show the error behavior?

Comment (MaxGekk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/IntervalUtilsSuite.scala]: Just to be clear, this PR just truncates results according to specified bound. In particular, it takes into account the left bound. Could you explain, please, what do you mean by saying `error behavior`?

Comment (cloud-fan) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/IntervalUtilsSuite.scala]: I think the conclusion is to fail for invalid bounds?

Comment (MaxGekk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/IntervalUtilsSuite.scala]: Do you mean if a string doesn't match to specified bounds like in `literals.sql`: https://github.com/apache/spark/pull/26358/files#diff-4f9e28af8e9fcb40a8a99b4e49f3b9b2R424 ?

Comment (cloud-fan) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/IntervalUtilsSuite.scala]: yea, like what Oracle does.

Comment (MaxGekk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/IntervalUtilsSuite.scala]: Here is the PR with strict parsing https://github.com/apache/spark/pull/26473

**Pull request closed without merging**